File: Geeks Island.py
class Solution():
    def isValid(self,r,c, n,m):
        if r>=n or r<0 or c>=m or c<0:
            return False
        return True
    def dfs(self, prev, r, c, mat, vis, n, m):
        if not self.isValid(r,c,n,m):
            return
        if vis[r][c]:
            return
        if mat[r][c] < prev:
            return
        vis[r][c]=True
        self.dfs(mat[r][c], r+1, c, mat,vis, n,m)
        self.dfs(mat[r][c], r-1, c, mat,vis, n,m)
        self.dfs(mat[r][c], r, c-1, mat,vis, n,m)
        self.dfs(mat[r][c], r, c+1, mat,vis, n,m)


    def water_flow(self, mat, n, m):
        indian = [[False for i in range(m)] for i in range(n)]
        arabian = [[False for i in range(m)] for i in range(n)]
        self.dfs(-1,0,4,mat,indian,n, m)
        for i in range(n): self.dfs(-1, i, 0, mat, indian, n, m)
        for i in range(m): self.dfs(-1, 0, i, mat, indian, n, m)
        for i in range(n): self.dfs(-1, i, m-1, mat, arabian, n, m)
        for i in range(m): self.dfs(-1, n-1, i, mat, arabian, n, m)
        ct  = 0
        for i in range(n):

            for j in range(m):
                if indian[i][j] and arabian[i][j]:
                    ct += 1
        return ct

File: Geeks Island.cpp
class Solution{   
public:
    void dfs(vector<vector<int>> &mat, vector<vector<bool>> &vis, int i, int j, int n, int m){

        if(i<0 || j<0 || i>=n || j>=m || vis[i][j]) return;
        vis[i][j] = true;
    
        if(i>0 && !vis[i-1][j] && mat[i-1][j] >= mat[i][j]){
            dfs(mat, vis, i-1, j, n, m);
        }
        if(i<n-1 && !vis[i+1][j] && mat[i+1][j] >= mat[i][j]){
            dfs(mat, vis, i+1, j, n, m);
        } 
        if(j>0 && !vis[i][j-1] && mat[i][j-1] >= mat[i][j]){
            dfs(mat, vis, i, j-1, n, m);
        } 
        if(j<m-1 && !vis[i][j+1] && mat[i][j+1] >= mat[i][j]){
            dfs(mat, vis, i, j+1, n, m);
        } 
    }

    int water_flow(vector<vector<int>> &mat,int N,int M){

        vector<vector<bool>> indian(N, vector<bool>(M, false));
        vector<vector<bool>> arabian(N, vector<bool>(M, false));
    
        for(int i=0;i<N;i++) dfs(mat, indian, i, 0, N, M);
        for(int i=0;i<M;i++) dfs(mat, indian, 0, i, N, M);
        for(int i=0;i<N;i++) dfs(mat, arabian, i, M-1, N, M);
        for(int i=0;i<M;i++) dfs(mat, arabian, N-1, i, N, M);
    
        int count = 0;
        for(int i=0;i<N;i++){
            for(int j=0;j<M;j++){
                if(indian[i][j] && arabian[i][j]){
                    count++;
                }        
            }
        }
    
        return count;
    }
};
